File: python/백준 온라인 저지/2022.02.10/기타 레슨.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
arr = list(map(int, input().split()))

start, end = max(arr), sum(arr)

while start <= end:
    mid = (start+end) // 2

    cnt = 0
    pt = 0
    for i in range(n):
        if pt + arr[i] > mid:
            cnt += 1
            pt = 0
        pt += arr[i]
    
    cnt += 1 if pt else 0

    if cnt <= m:
        end = mid - 1
    else:
        start = mid + 1

print(start)

File: python/백준 온라인 저지/2022.02.10/기타 레슨.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.collections.maxOrNull as maxOrNull1

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val arr = readLine().split(" ").map { it.toInt() }.toIntArray()

    var start = arr.maxOrNull1()!!
    var end = arr.sum()!!

    var mid: Int
    var cnt: Int
    var pt: Int

    while (start <= end) {
        mid = (start + end) / 2
        cnt = 0
        pt = 0
        for (i in 0 until n) {
            when {
                pt + arr[i] > mid -> {
                    cnt += 1
                    pt = 0
                }
            }
            pt += arr[i]
        }

        cnt += if (pt != 0) 1 else 0
        when {
            cnt <= m -> {
                end = mid - 1
            }
            else -> {
                start = mid + 1
            }
        }
    }
    println(start)
}